File: javascript/sails/api/controllers/UserController.js
/**
 * UserController
 *
 * @description :: Server-side actions for handling user requests.
 */

module.exports = {
    /**
     * Get all users
     */
    getAllUsers: async function(req, res) {
      try {
        // Using require to directly access the service
        const UserService = require('../services/UserService');
        const users = await UserService.getAllUsers();
        return res.json(users);
      } catch (error) {
        return res.serverError(error);
      }
    },
  
    /**
     * Get user by ID
     */
    getUserById: async function(req, res) {
      try {
        const userId = req.params.id;
        const UserService = require('../services/UserService');
        const user = await UserService.getUserById(userId);
        return res.json(user);
      } catch (error) {
        if (error.status === 404) {
          return res.notFound(error);
        }
        return res.serverError(error);
      }
    },
  
    /**
     * Create a new user
     */
    createUser: async function(req, res) {
      try {
        const userData = req.body;
        const UserService = require('../services/UserService');
        const newUser = await UserService.createUser(userData);
        return res.status(201).json(newUser);
      } catch (error) {
        return res.serverError(error);
      }
    },
  
    /**
     * Update an existing user
     */
    updateUser: async function(req, res) {
      try {
        const userId = req.params.id;
        const userData = req.body;
        const UserService = require('../services/UserService');
        const updatedUser = await UserService.updateUser(userId, userData);
        return res.json(updatedUser);
      } catch (error) {
        if (error.status === 404) {
          return res.notFound(error);
        }
        return res.serverError(error);
      }
    },
  
    /**
     * Delete a user
     */
    deleteUser: async function(req, res) {
      try {
        const userId = req.params.id;
        const UserService = require('../services/UserService');
        const deletedUser = await UserService.deleteUser(userId);
        return res.json(deletedUser);
      } catch (error) {
        if (error.status === 404) {
          return res.notFound(error);
        }
        return res.serverError(error);
      }
    }
  };

File: otel-collector/otel-operator/java/src/main/java/com/example/controller/UserController.java
package com.example.controller;

import com.example.model.User;
import com.example.service.UserService;
import jakarta.validation.Valid;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

/**
 * REST Controller for User operations
 */
@RestController
@RequestMapping("/api/users")
@CrossOrigin(origins = "*")
public class UserController {

    private static final Logger log = LoggerFactory.getLogger(UserController.class);
    
    private final UserService userService;

    @Autowired
    public UserController(UserService userService) {
        this.userService = userService;
    }

    /**
     * Create a new user
     */
    @PostMapping
    public ResponseEntity<User> createUser(@Valid @RequestBody User user) {
        try {
            User createdUser = userService.createUser(user);
            log.info("User created successfully: {}", createdUser.getUsername());
            return ResponseEntity.status(HttpStatus.CREATED).body(createdUser);
        } catch (RuntimeException e) {
            log.error("Error creating user: {}", e.getMessage());
            return ResponseEntity.badRequest().build();
        }
    }

    /**
     * Get all users
     */
    @GetMapping
    public ResponseEntity<List<User>> getAllUsers() {
        try {
            List<User> users = userService.getAllUsers();
            log.info("Retrieved {} users", users.size());
            return ResponseEntity.ok(users);
        } catch (Exception e) {
            log.error("Error retrieving users: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * Get user by ID
     */
    @GetMapping("/{id}")
    public ResponseEntity<User> getUserById(@PathVariable Long id) {
        try {
            Optional<User> user = userService.getUserById(id);
            if (user.isPresent()) {
                log.info("User found with ID: {}", id);
                return ResponseEntity.ok(user.get());
            } else {
                log.warn("User not found with ID: {}", id);
                return ResponseEntity.notFound().build();
            }
        } catch (Exception e) {
            log.error("Error retrieving user with ID {}: {}", id, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * Get user by username
     */
    @GetMapping("/username/{username}")
    public ResponseEntity<User> getUserByUsername(@PathVariable String username) {
        try {
            Optional<User> user = userService.getUserByUsername(username);
            if (user.isPresent()) {
                log.info("User found with username: {}", username);
                return ResponseEntity.ok(user.get());
            } else {
                log.warn("User not found with username: {}", username);
                return ResponseEntity.notFound().build();
            }
        } catch (Exception e) {
            log.error("Error retrieving user with username {}: {}", username, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * Update user
     */
    @PutMapping("/{id}")
    public ResponseEntity<User> updateUser(@PathVariable Long id, @Valid @RequestBody User userDetails) {
        try {
            User updatedUser = userService.updateUser(id, userDetails);
            log.info("User updated successfully with ID: {}", id);
            return ResponseEntity.ok(updatedUser);
        } catch (RuntimeException e) {
            log.error("Error updating user with ID {}: {}", id, e.getMessage());
            return ResponseEntity.badRequest().build();
        } catch (Exception e) {
            log.error("Error updating user with ID {}: {}", id, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * Delete user
     */
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteUser(@PathVariable Long id) {
        try {
            userService.deleteUser(id);
            log.info("User deleted successfully with ID: {}", id);
            return ResponseEntity.noContent().build();
        } catch (RuntimeException e) {
            log.error("Error deleting user with ID {}: {}", id, e.getMessage());
            return ResponseEntity.badRequest().build();
        } catch (Exception e) {
            log.error("Error deleting user with ID {}: {}", id, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * Search users by name
     */
    @GetMapping("/search")
    public ResponseEntity<List<User>> searchUsersByName(@RequestParam String name) {
        try {
            List<User> users = userService.searchUsersByName(name);
            log.info("Found {} users matching name: {}", users.size(), name);
            return ResponseEntity.ok(users);
        } catch (Exception e) {
            log.error("Error searching users by name {}: {}", name, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * Get users by first name
     */
    @GetMapping("/firstname/{firstName}")
    public ResponseEntity<List<User>> getUsersByFirstName(@PathVariable String firstName) {
        try {
            List<User> users = userService.getUsersByFirstName(firstName);
            log.info("Found {} users with first name: {}", users.size(), firstName);
            return ResponseEntity.ok(users);
        } catch (Exception e) {
            log.error("Error retrieving users by first name {}: {}", firstName, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * Get users by last name
     */
    @GetMapping("/lastname/{lastName}")
    public ResponseEntity<List<User>> getUsersByLastName(@PathVariable String lastName) {
        try {
            List<User> users = userService.getUsersByLastName(lastName);
            log.info("Found {} users with last name: {}", users.size(), lastName);
            return ResponseEntity.ok(users);
        } catch (Exception e) {
            log.error("Error retrieving users by last name {}: {}", lastName, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    /**
     * Initialize sample data
     */
    @PostMapping("/init")
    public ResponseEntity<String> initializeData() {
        try {
            userService.initializeSampleData();
            return ResponseEntity.ok("Sample data initialized successfully!");
        } catch (Exception e) {
            log.error("Error initializing sample data: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Error initializing data");
        }
    }

    /**
     * Health check endpoint
     */
    @GetMapping("/health")
    public ResponseEntity<String> health() {
        return ResponseEntity.ok("User service is running!");
    }
} 